docs/adr/adr-003-march2022-testnet.md

```diff
@@ -0,0 +1,159 @@
+to `TxSub` and relaying the transactions into the Core mempool.
+
+Celestia-node's state interaction will be detailed further in a subsequent ADR.
+
+### [Data Availability Sampling during `HeaderSync`](https://github.com/celestiaorg/celestia-node/issues/181)
+
+Currently, both **light** and **full* nodes are unable to perform data availability sampling (DAS) while syncing.
+They only begin sampling once the node is synced up to head of chain. 
+
+`HeaderSync` and the `DASer` will be refactored such that the `DASer` will be able to perform sampling on past headers
+as the node is syncing. To do this, the syncing algorithms in both the `DASer` and `HeaderSync` should align so that
+headers received during sync will be propagated to the `DASer` for sampling via an internal pubsub.
+
+The `DASer` will maintain a checkpoint to the last sampled header so that it can continue sampling from the last 
+checkpoint on any new headers.
+
+
+<hr style="border:1px solid gray"> </hr>
+
+## Refactoring
+
+### `HeaderService` becomes main component around which most other services are focused
+Initially, we started with BlockService being the more “important” component during devnet architecture, but overlooked
+some problems with regards to sync (we initially made the decision that a celestia full node would have to be started 
+together at the same time as a core node -- which is the reason for embedding the core node).
+
+This led us to an issue where eventually we needed to connect to an already-running core node and sync from it. We were
+missing a component to do that, so we implemented `HeaderExchange` over the core client (wrapping another interface we 
+had previously created for BlockService called `BlockFetcher`), and we had to do this last minute b/c it wouldn’t work 
+otherwise, leading to a bunch of hacks and other issues (like having to hand the celestia full node a “trusted” hash of
+a header from the already-running chain so that it can sync up to that point and start listening for new headers.
+
+**Proposed new architecture**:
+
+### [`BlockService` is only responsible for reconstructing the block from Shares handed to it by the `ShareService`](https://github.com/celestiaorg/celestia-node/issues/251).
+Right now, the `BlockService` is in charge of fetching new blocks from the core node, erasure coding them, generating 
+DAH, generating `ExtendedHeader`, broadcasting `ExtendedHeader` to `HeaderSub` network, and storing the block data 
+(after some validation checks).
+
+Instead, we should rely on ShareService sampling to fetch us *enough* shares to reconstruct the block inside of 
+`BlockService`.
+
+### `ShareService` optimizations
+* Implement parallelization for retrieving shares by namespace. This
+[issue](https://github.com/celestiaorg/celestia-node/issues/184) is already being worked on.
+* NMT/Shares/Namespace storage optimizations (**TODO @WONDERTAN**)
+* Pruning/GC for shares.(**TODO @WONDERTAN**)
```

Comment:
Hmm, does it make sense for light nodes to keep shares only for headers within unbonding period and not beyond? Same for headers, a light node may simply delete all the past headers beyond unbonding period, as it does not trust them anyway.

Reply:
Technically the same is true for Full Node as they are also prone to longe-range attack(because they don't do consensus validation), but someone still have to serve all the headers, so they should keep them.

Reply:
Extracted into https://github.com/celestiaorg/celestia-node/issues/272